File: web/src/services/tracker/TrackerService.js
import { request, METHOD } from "@/utils/request";
export function findTrackers(parameter) {
  return request({
    url: "tracker/project/tracker/list",
    method: METHOD.GET,
    params: parameter,
  });
}
export function findOneTracker(trackerId) {
  return request({
    url: `tracker/project/tracker/${trackerId}`,
    method: METHOD.GET,
  });
}

export function findTrackerByIds(trackerIds) {
  // 根据 trackerIds 查找 tracker
  if (trackerIds && trackerIds.length > 0) {
    return request({
      url:
        `tracker/project/tracker/findByIds?` +
        trackerIds.map((_) => `trackerIds=${_}`).join("&"),
      method: METHOD.GET,
    });
  } else {
    // 如果 trackerIds 为空或长度为0，直接返回一个空数组的 Promise
    return Promise.resolve([]);
  }
}

export function createTracker(parameter) {
  return request({
    url: "tracker/project/tracker",
    method: METHOD.POST,
    data: parameter,
  });
}

export function updateTracker(trackerId, parameter) {
  return request({
    url: `tracker/project/tracker/${trackerId}`,
    method: METHOD.PUT,
    data: parameter,
  });
}
export function deleteTracker(trackerId) {
  return request({
    url: `tracker/project/tracker/${trackerId}`,
    method: METHOD.DELETE,
  });
}

export function updateTrackerOrdinary(trackers) {
  return request({
    url: `tracker/project/tracker/sort`,
    method: METHOD.POST,
    data: trackers,
  });
}

export function copyTracker(tracker) {
  return request({
    url: `tracker/project/tracker/copy`,
    method: METHOD.POST,
    data: tracker,
  });
}
export function copyTrackerFields(trackerId, fields) {
  return request({
    url: `tracker/project/tracker/${trackerId}/copy-fields`,
    method: METHOD.POST,
    data: fields,
  });
}

export function findTrackerFields(trackerId) {
  return request({
    url: `/tracker/project/tracker/${trackerId}/fields`,
    method: METHOD.GET,
  });
}
export function createTrackerField(trackerId, field) {
  return request({
    url: `/tracker/project/tracker/${trackerId}/fields`,
    method: METHOD.POST,
    data: field,
  });
}
export function updateTrackerField(trackerId, field) {
  return request({
    url: `/tracker/project/tracker/${trackerId}/fields`,
    method: METHOD.PUT,
    data: field,
  });
}
export function deleteTrackerField(trackerId, field) {
  return request({
    url: `/tracker/project/tracker/${trackerId}/fields`,
    method: METHOD.DELETE,
    data: field,
  });
}

export function updateTrackerLayout(trackerId, layout) {
  return request({
    url: `tracker/project/tracker/${trackerId}/updateLayout`,
    method: METHOD.PUT,
    data: layout,
  });
}

export function updateTrackerPermission(trackerId, rolePermission) {
  return request({
    url: `tracker/project/tracker/${trackerId}/updatePermission`,
    method: METHOD.PUT,
    data: rolePermission,
  });
}

export function findTrackerLayout(trackerId, layoutName) {
  return request({
    url: `tracker/project/tracker/${trackerId}/findLayout`,
    method: METHOD.GET,
    params: { layoutName: layoutName },
  });
}

export function createTrackerStatus(trackerId, trackerStatus) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status`,
    method: METHOD.POST,
    data: trackerStatus,
  });
}
export function updateTrackerStatus(trackerId, trackerStatus) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status`,
    method: METHOD.PUT,
    data: trackerStatus,
  });
}
export function deleteTrackerStatus(trackerId, trackerStatus) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status`,
    method: METHOD.DELETE,
    data: trackerStatus,
  });
}

export function findTrackerStatus(trackerId) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status`,
    method: METHOD.GET,
  });
}

export function sortTrackerStatus(trackerId, trackerStatusList) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status/sort`,
    method: METHOD.POST,
    data: trackerStatusList,
  });
}

export function setInitialTrackerStatus(trackerId, trackerStatus) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status/setInitial`,
    method: METHOD.POST,
    data: trackerStatus,
  });
}

export function trackerStatusUseByTransition(trackerId, trackerStatus) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status/useByTransition`,
    method: METHOD.POST,
    data: trackerStatus,
  });
}

export function trackerStatusUseByWorkItem(trackerId, trackerStatus) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status/useByWorkItem`,
    method: METHOD.POST,
    data: trackerStatus,
  });
}

export function replaceWorkItemStatus(
  trackerId,
  oldTrackerStatusId,
  newTrackerStatusId
) {
  return request({
    url: `tracker/project/tracker/${trackerId}/status/replaceWorkItemStatus`,
    method: METHOD.POST,
    params: {
      oldTrackerStatusId: oldTrackerStatusId,
      newTrackerStatusId: newTrackerStatusId,
    },
  });
}

export function updateTrackerStateTransitions(trackerId, transitions) {
  return request({
    url: `tracker/project/tracker/${trackerId}/stateTransitions`,
    method: METHOD.PUT,
    data: transitions,
  });
}

export function findTrackerStateTransitions(trackerId) {
  return request({
    url: `tracker/project/tracker/${trackerId}/stateTransitions`,
    method: METHOD.GET,
  });
}

export function createTrackerStateTransition(trackerId, stateTransition) {
  return request({
    url: `tracker/project/tracker/${trackerId}/transition`,
    method: METHOD.POST,
    data: stateTransition,
  });
}
export function updateTrackerStateTransition(trackerId, stateTransition) {
  return request({
    url: `tracker/project/tracker/${trackerId}/transition`,
    method: METHOD.PUT,
    data: stateTransition,
  });
}
export function deleteTrackerStateTransition(trackerId, stateTransition) {
  return request({
    url: `tracker/project/tracker/${trackerId}/transition`,
    method: METHOD.DELETE,
    data: stateTransition,
  });
}

export function renameTrackerStateTransition(trackerId, stateTransition) {
  return request({
    url: `tracker/project/tracker/${trackerId}/transition/rename`,
    method: METHOD.POST,
    data: stateTransition,
  });
}
export function grantTrackerStateTransitionPermitted(
  trackerId,
  transitionId,
  identity
) {
  return request({
    url: `tracker/project/tracker/${trackerId}/transition/grant`,
    method: METHOD.POST,
    params: { transitionId: transitionId },
    data: identity,
  });
}
export function unGrantTrackerStateTransitionPermitted(
  trackerId,
  transitionId,
  identity
) {
  return request({
    url: `tracker/project/tracker/${trackerId}/transition/unGrant`,
    method: METHOD.POST,
    params: { transitionId: transitionId },
    data: identity,
  });
}

export function findTrackerPermissions(trackerId) {
  return request({
    url: `tracker/project/tracker/${trackerId}/permissions`,
    method: METHOD.GET,
  });
}

export function findTrackerPermissionsDefault() {
  return request({
    url: `tracker/project/tracker/permissions/default`,
    method: METHOD.GET,
  });
}

export function grantTrackerPermission(trackerId, permission, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/permissions/grant`,
    method: METHOD.POST,
    params: { permissionName: permission },
    data: identity,
  });
}
export function unGrantTrackerPermission(trackerId, permission, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/permissions/unGrant`,
    method: METHOD.POST,
    params: { permissionName: permission },
    data: identity,
  });
}

export function findSystemEventDefault() {
  return request({
    url: `tracker/project/tracker/notifications/system/default`,
    method: METHOD.GET,
  });
}

export function subscribeSystemEvent(trackerId, eventName, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/system/subscribe`,
    method: METHOD.POST,
    params: { eventName: eventName },
    data: identity,
  });
}
export function unSubscribeSystemEvent(trackerId, eventName, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/system/unsubscribe`,
    method: METHOD.POST,
    params: { eventName: eventName },
    data: identity,
  });
}
export function subscribeCustomerEvent(trackerId, fieldId, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/subscribe`,
    method: METHOD.POST,
    params: { fieldId: fieldId },
    data: identity,
  });
}
export function unSubscribeCustomerEvent(trackerId, fieldId, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/unsubscribe`,
    method: METHOD.POST,
    params: { fieldId: fieldId },
    data: identity,
  });
}
export function subscribeAllCustomerEvent(trackerId, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/all/subscribe`,
    method: METHOD.POST,
    data: identity,
  });
}
export function unSubscribeAllCustomerEvent(trackerId, identity) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/all/unsubscribe`,
    method: METHOD.POST,
    data: identity,
  });
}
export function createCustomerNotification(trackerId, notification) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/`,
    method: METHOD.POST,
    data: notification,
  });
}
export function deleteCustomerNotification(trackerId, notification) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/`,
    method: METHOD.DELETE,
    data: notification,
  });
}

export function updateSystemEventNoticeType(
  trackerId,
  eventName,
  notification
) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/system/noticeType`,
    method: METHOD.POST,
    params: { eventName: eventName },
    data: notification,
  });
}
export function updateCustomerEventNoticeType(
  trackerId,
  fieldId,
  notification
) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/noticeType`,
    method: METHOD.POST,
    params: { fieldId: fieldId },
    data: notification,
  });
}
export function updateAllCustomerEventNoticeType(trackerId, notification) {
  return request({
    url: `tracker/project/tracker/${trackerId}/notifications/customer/all/noticeType`,
    method: METHOD.POST,
    data: notification,
  });
}
export function updateTrackerName(trackerId, tracker) {
  return request({
    url: `tracker/project/tracker/${trackerId}/nameAndType`,
    method: METHOD.PUT,
    data: tracker,
  });
}

File: server/helm-tracker/src/main/java/com/harbortek/helm/tracker/service/TrackerService.java
/*
 * Copyright [2025] [Harbortek Corp.]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.harbortek.helm.tracker.service;

import com.harbortek.helm.common.vo.BaseIdentity;
import com.harbortek.helm.system.vo.TrackerFieldsPermissions;
import com.harbortek.helm.tracker.entity.tracker.TrackerEntity;
import com.harbortek.helm.tracker.vo.permission.PermissionGrantVo;
import com.harbortek.helm.tracker.vo.tracker.TrackerVo;
import com.harbortek.helm.tracker.vo.tracker.fields.TrackerField;
import com.harbortek.helm.tracker.vo.tracker.layout.TrackerLayout;
import com.harbortek.helm.tracker.vo.tracker.nofitication.CustomerTrackerNotification;
import com.harbortek.helm.tracker.vo.tracker.nofitication.SystemTrackerNotification;
import com.harbortek.helm.tracker.vo.tracker.nofitication.TrackerNotification;
import com.harbortek.helm.tracker.vo.tracker.permissions.TrackerRolePermission;
import com.harbortek.helm.tracker.vo.tracker.stateTransition.TrackerStateTransition;
import com.harbortek.helm.tracker.vo.tracker.stateTransition.TrackerStatus;

import java.util.Collection;
import java.util.List;

public interface TrackerService {

	TrackerVo createTracker(TrackerVo tracker);

	TrackerVo updateTracker(TrackerVo tracker);

	void deleteTracker(TrackerVo tracker);


	TrackerVo findOneTracker(Long id);

	List<TrackerVo> findByProject(Long projectId);

	List<TrackerVo> findByProject(Long projectId, boolean showInternal);

	List<TrackerVo> findTrackerByIds(List<Long> trackerIds);

	void updateTrackerLayout(Long trackerId, TrackerLayout trackerLayout);

	void fillTrackerFields(List<TrackerField> trackerFields, List<TrackerStatus> trackerStatuses, Long projectId);

//	void updateTrackerPermission(Long trackerId, TrackerRolePermission trackerRolePermission);


	TrackerLayout findOneTrackerLayout(Long trackerId, String layoutName);

	List<TrackerStateTransition> findTrackerStateTransitions(Long trackerId);

	void updateTrackerStateTransitions(Long trackerId, List<TrackerStateTransition> trackerStateTransitions);

	void updateTrackerStatus(Long trackerId, TrackerStatus trackerStatus);

	void deleteTrackerStatus(Long trackerId, TrackerStatus trackerStatus);

	List<TrackerStatus> findTrackerStatus(Long trackerId);

	TrackerStatus createTrackerStatus(Long trackerId, TrackerStatus trackerStatus);

	void sortTrackerStatus(Long trackerId, List<TrackerStatus> trackerStatuses);

	void setInitialTrackerStatus(Long trackerId, TrackerStatus trackerStatus);

	boolean trackerStatusUseByTransition(Long trackerId, TrackerStatus trackerStatus);

	boolean trackerStatusUseByWorkItem(Long trackerId, TrackerStatus trackerStatus);

	TrackerStateTransition createTrackerStateTransition(Long trackerId, TrackerStateTransition trackerStateTransition);

	void updateTrackerStateTransition(Long trackerId, TrackerStateTransition trackerStateTransition);

	void deleteTrackerStateTransition(Long trackerId, TrackerStateTransition trackerStateTransition);

	void renameTrackerStateTransition(Long trackerId, TrackerStateTransition trackerStateTransition);

    List<TrackerField> findTrackerFields(Long trackerId);

	TrackerField createTrackerField(Long trackerId, TrackerField trackerField);

	void updateTrackerField(Long trackerId, TrackerField trackerField);

	void deleteTrackerField(Long trackerId, TrackerField trackerField);

    List<PermissionGrantVo> findTrackerPermissions(Long trackerId);

	void grantTrackerPermission(Long trackerId, String permissionName, BaseIdentity identity);

	void unGrantTrackerPermission(Long trackerId, String permissionName, BaseIdentity identity);


	TrackerNotification findTrackerNotification(Long trackerId);

	void subscribeSystemEvent(Long trackerId, String eventName, BaseIdentity identity);

	void unSubscribeSystemEvent(Long trackerId, String eventName, BaseIdentity identity);

	void subscribeCustomerEvent(Long trackerId, Long fieldId, BaseIdentity identity);

	void unSubscribeCustomerEvent(Long trackerId, Long fieldId, BaseIdentity identity);

	void createCustomerNotification(Long trackerId, CustomerTrackerNotification notification);

	void deleteCustomerNotification(Long trackerId, CustomerTrackerNotification notification);

	void subscribeAllCustomerEvent(Long trackerId, BaseIdentity identity);

	void unSubscribeAllCustomerEvent(Long trackerId, BaseIdentity identity);

	void updateSystemEventNoticeType(Long trackerId, String eventName, SystemTrackerNotification systemTrackerNotification);

	void updateCustomerEventNoticeType(Long trackerId, Long fieldId, CustomerTrackerNotification customerTrackerNotification);

	void updateAllCustomerEventNoticeType(Long trackerId, CustomerTrackerNotification customerTrackerNotification);

	void grantTrackerStateTransitionPermitted(Long trackerId, Long transitionId, BaseIdentity identity);

	void unGrantTrackerStateTransitionPermitted(Long trackerId, Long transitionId, BaseIdentity identity);

    void updateTrackerOrdinary(List<TrackerVo> trackerVos);

	void copyTracker(TrackerVo trackerVo);

	void copyTrackerFields(Long trackerId, List<TrackerField> trackerFields);

	void updateTrackerNameAndType(TrackerVo trackerVo);

	void replaceWorkItemStatus(Long trackerId, Long oldTrackerStatusId, Long newTrackerStatusId);

    void createTrackers(Collection<TrackerVo> trackers);

    TrackerVo findOneTrackerByName(Long projectId, String trackerName);
}
